$\newcommand{\xv}{\mathbf{x}}
 \newcommand{\wv}{\mathbf{w}}
 \newcommand{\yv}{\mathbf{y}}
 \newcommand{\zv}{\mathbf{z}}
 \newcommand{\uv}{\mathbf{u}}
 \newcommand{\vv}{\mathbf{v}}
 \newcommand{\Chi}{\mathcal{X}}
 \newcommand{\R}{\rm I\!R}
 \newcommand{\sign}{\text{sign}}
 \newcommand{\Tm}{\mathbf{T}}
 \newcommand{\Xm}{\mathbf{X}}
 \newcommand{\Zm}{\mathbf{Z}}
 \newcommand{\I}{\mathbf{I}}
 \newcommand{\Um}{\mathbf{U}}
 \newcommand{\Vm}{\mathbf{V}}
 \newcommand{\muv}{\boldsymbol\mu}
 \newcommand{\Sigmav}{\boldsymbol\Sigma}
 \newcommand{\Lambdav}{\boldsymbol\Lambda}
$

# Text Data Classification


### ITCS 5156
### Minwoo "Jake" Lee

<br/>
<font color="blue"><b>

NAME: Indhuja Gudluru

</b> </font>

# Goal

The goal of this activity is to be familar to the preprocessing of text data and its classification afterwards. We will use the built-in 20 Newsgroups dataset in Scikit-Learn, which has approximately 20,000 newsgroup documents. Follow the TODO titles and comments to finish the activity!

# Agenda

* Loading Newsgroup data from Scikit-Learn
* Preprocessing & Classificaiton with
  * Back of Words
  * TF-IDF
  * n-Gram
  * Word2Vec
  * RoBERTa


# Tables of TODO's


1. [TODO1 (5 points)](#TODO1)
2. [TODO2 (5 points)](#TODO2)
3. [TODO3 (5 points)](#TODO3)
4. [TODO4 (5 points)](#TODO4)  
5. [TODO5 (5 points)](#TODO5)
6. [TODO6 (5 points)](#TODO6)
7. [TODO7 (5 points)](#TODO7)
8. [TODO8 (5 points)](#TODO8)
9. [TODO9 (20 points)](#TODO9)
10. [TODO10 (18 points)](#TODO10)
11. [TODO11 (20 points)](#TODO11)

12. [Feedback (2 points)](#Feedback)


Well, let us start the lab with importing the common libraries first.

In [1]:
from copy import deepcopy as copy

import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
from mpl_toolkits.mplot3d import Axes3D
%matplotlib inline

import seaborn as sns

# Loading 20 Newsgroup data from Scikit-Learn

Let us start the practice by loading the newgroup data.
The news group classes are as the table below:

<img src="https://d3i71xaburhd42.cloudfront.net/6eb6a45225d7d6a9ab0bb055b81d9aeaba867930/3-Table1-1.png" width=500 />

To save the computation time, let us selct the five categories as given in the codes below.

<div id="TODO1"></div>

### TODO1 (5 points)
1. Use `fetch_20newsgroups` to load train data using the `subset` option and store to `train_data`. Don't forget to pass the categories to the function.
2. Using the same function, load test data and store it to `test_data`.

In [2]:
from sklearn.datasets import fetch_20newsgroups

cat = ['alt.atheism', 'soc.religion.christian',
       'comp.graphics', 'sci.med', 'rec.sport.baseball']

# 1)
train_data = fetch_20newsgroups(subset='train', categories=cat)


# 2)
test_data = fetch_20newsgroups(subset='test', categories=cat)




In [3]:
train_data.target_names

['alt.atheism',
 'comp.graphics',
 'rec.sport.baseball',
 'sci.med',
 'soc.religion.christian']

In [4]:
train_data

{'data': ["From: harvey@oasys.dt.navy.mil (Betty Harvey)\nSubject: Re: Is MSG sensitivity superstition?\nReply-To: harvey@oasys.dt.navy.mil (Betty Harvey)\nOrganization: Carderock Division, NSWC, Bethesda, MD\nLines: 30\n\nIn rec.food.cooking, packer@delphi.gsfc.nasa.gov (Charles Packer) writes:\n>Is there such a thing as MSG (monosodium glutamate) sensitivity?\n>I saw in the NY Times Sunday that scientists have testified before\n>an FDA advisory panel that complaints about MSG sensitivity are\n>superstition. Anybody here have experience to the contrary?\n>\nI know that there is MSG sensitivity.  When I eat foods with MSG I get\nvery thirsty and my hands swell and get a terrible itchy rash. I first\nexperienced this problem when I worked close to Chinatown and ate Chinese\nfood almost everyday for lunch.  Now I can't tolerate MSG at all.  I can\nnotice immediately when I have eaten any.  I try to avoid MSG completely.\n\nInteresting fact though is that all three of my children started 

In [5]:
test_data

{'data': ["Organization: University of Maine System\nFrom: Ryan Robbins <IO20456@MAINE.MAINE.EDU>\nSubject: Re: lame, dumb and useless question\nLines: 10\n\nIt depends. If, in the judgment of the umpire the batter made no\nattempt to avoid getting hit, the batter is awarded first for a\nbase on balls. If the umpire rules he did try to get out of the\nway, he's awarded first because of a hit batsman.\n\nRyan Robbins\nPenobscot Hall\nUniversity of Maine\n\nIO20456@Maine.Maine.Edu\n",
  'From: ez019654@othello.ucdavis.edu (Victoria Milliron)\nSubject: Intel\'s PCI standard???\nOrganization: University of California, Davis\nX-Newsreader: Tin 1.1 PL3\nLines: 8\n\n       I read a mesg. somewhere on GENIE about Intel coming out with a \ngraphics standard called PCI, which would supplant VESA standards.  Is\nthis a rumor, or is there some substance to it. If any of y\'all have\nheard of this "standard" please e-mail me on how I might obtain more info\n\nThanks in Advance\nvamilliron@othello.u

In [6]:
len(train_data.data), len(test_data.data)

(2854, 1899)

# Back of Words

We now preprocess the text data with back-of-words processing, which includes tokenizing, filtering of stopwords, and encoding. It is all implemented in CountVectorizer, so let us play with it.



<div id="TODO2"></div>

### TODO2 (5 points)

1. Create vectorizer `vect` instance of `CountVectorizer` without any custom configuration.
2. By using `fit_transform`, transform the `train_data` into `X_train_cnt`.
3. Transform the `test_data` into `X_test_cnt`.


In [7]:
from sklearn.feature_extraction.text import CountVectorizer

# 1)
vect = CountVectorizer()
# 2)

X_train_cnt = vect.fit_transform(train_data.data)

In [8]:
X_train_cnt.shape

(2854, 39806)

In [9]:
# 3)

X_test_cnt = vect.transform(test_data.data)

## Classificaiton

<div id="TODO3"></div>

### TODO3 (5 points)

1. Create a Naive Bayes classifer instance `nbc` with `MultinomialNB`.
2. Train the classifier with transformed feature `X_train_cnt` and the label `train_data.target`.
3. Make a prediction for the test features `X_test_cnt`.
4. Compute the f1-score to measure the performance.

In [10]:
from sklearn.naive_bayes import MultinomialNB

# 1)
nbc = MultinomialNB()
# 2)
nbc.fit(X_train_cnt, train_data.target)
# 3)
pred = nbc.predict(X_test_cnt)

pred

array([2, 1, 2, ..., 3, 3, 4])

In [11]:
# 4)
from sklearn.metrics import f1_score
f1Score = f1_score(test_data.target, pred, average='weighted')
f1Score

0.9376965803744591

<div id="TODO4"></div>

### TODO4 (5 points)

To make this easy to work with, let us build a pipeline of vectorizer and a classifier.

1. Create a `Pipeline` instance that connects `CountVectorizer` and `MultinomialNB` and store it to `nbclf`.
2. Train `nbclf` with train_data. Use the attributes `data` and `target`.
3. Make a prediciton and store the results in `pred`.
4. Compute the f1 score with `test_data`.

In [12]:
from sklearn.pipeline import Pipeline

# TODO 4.1 - 4.4

nbclf = Pipeline([
    ('vect', CountVectorizer()),
    ('clf', MultinomialNB()),
])
nbclf.fit(train_data.data, train_data.target)
pred = nbclf.predict(test_data.data)
f1Score = f1_score(test_data.target, pred, average='weighted')
f1Score


0.9376965803744591

We have very good first results. As the text contains a lot of meaningless words. Let us discard the stop words and see how the performance differs.  

<div id="TODO5"></div>

### TODO5 (5 points)

Repeat TODO4 with stop words this time.

1. Create a `Pipeline` instance `nbclf1`. Make sure to create CountVectorizer instance with `stop_words`.
2. Train `nbclf1` with train_data. Use the attributes `data` and `target`.
3. Make a prediciton and store the results in `pred`.
4. Print out the `classification_report` with `test_data`.


In [13]:
# TODO 5.1 - 5.4
from sklearn.metrics import classification_report
nbclf1 = Pipeline([
    ('vect', CountVectorizer(stop_words='english')),
    ('clf', MultinomialNB()),
])
nbclf1.fit(train_data.data, train_data.target)
pred = nbclf1.predict(test_data.data)
print(classification_report(test_data.target, pred, target_names=test_data.target_names))




                        precision    recall  f1-score   support

           alt.atheism       0.93      0.90      0.91       319
         comp.graphics       0.94      0.95      0.95       389
    rec.sport.baseball       0.97      0.98      0.98       397
               sci.med       0.96      0.91      0.93       396
soc.religion.christian       0.93      0.96      0.94       398

              accuracy                           0.94      1899
             macro avg       0.94      0.94      0.94      1899
          weighted avg       0.94      0.94      0.94      1899




The f1 score has been slightly increased. Some words appear only once might not be a key component to describe the document. So, in this time let us see what happens if we ignore them.  

<div id="TODO6"></div>

### TODO6 (5 points)

Repeat TODO4 with `min_df` this time.

1. Create a `Pipeline` instance `nbclf2`. Make sure to create CountVectorizer instance with `min_df=2`.
2. Train `nbclf2` with train_data. Use the attributes `data` and `target`.
3. Make a prediciton and store the results in `pred`.
4. Print out the `classification_report` with `test_data`.


In [14]:
# TODO 6.1 - 6.4

nbclf2 = Pipeline([
    ('vect', CountVectorizer(min_df=2)),
    ('clf', MultinomialNB()),
])
nbclf2.fit(train_data.data, train_data.target)
pred = nbclf2.predict(test_data.data)
print(classification_report(test_data.target, pred, target_names=test_data.target_names))




                        precision    recall  f1-score   support

           alt.atheism       0.91      0.91      0.91       319
         comp.graphics       0.93      0.96      0.94       389
    rec.sport.baseball       0.97      0.98      0.97       397
               sci.med       0.96      0.90      0.93       396
soc.religion.christian       0.94      0.96      0.95       398

              accuracy                           0.94      1899
             macro avg       0.94      0.94      0.94      1899
          weighted avg       0.94      0.94      0.94      1899



The score is similar to removing stop words. That means encoding the count vector with the descriptive words is more effective is helpfult to improve classificaiton accuracy.

## TF-IDF

We can extract a feature representation by using the term frequency and inverse document frequency using the equaion:
$$
 tfidf(w, d) = tf \times \log \Big( \frac{N+1}{N_w + 1} \Big) + 1.
$$
Let us transform the count vector into tf-idf feature representation and see how it affects the performance.

<div id="TODO7"></div>

### TODO7 (5 points)

1. Create `TfidfTransformer` instance and name it `tf_transformer`.
2. `fit_transform` the `X_train_cnt`, store it in `tfidf`, and print the shape.

In [15]:
from sklearn.feature_extraction.text import TfidfTransformer

# TODO 7.1 - 7.2
tf_transformer = TfidfTransformer()
tfidf = tf_transformer.fit_transform(X_train_cnt)


tfidf.shape

(2854, 39806)

Now, let us insert this to the pipelne for `nbclf1` with stop words and see how it performs.

<div id="TODO8"></div>

### TODO8 (5 points)

1. Create a `Pipeline` instance `nbclf3` with `CountVectorizer`, `TfidfTransformer`, and `MultinomialNB`.
2. Train `nbclf3` with train_data. Use the attributes `data` and `target`.
3. Make a prediciton and store the results in `pred`.
4. Print out the `classification_report` with `test_data`.


In [16]:
# TODO 8.1 - 8.4

nbclf3 = Pipeline([
    ('vect', CountVectorizer()),
    ('tfidf', TfidfTransformer()),
    ('clf', MultinomialNB()),
])
nbclf3.fit(train_data.data, train_data.target)
pred = nbclf3.predict(test_data.data)
print(classification_report(test_data.target, pred, target_names=test_data.target_names))





                        precision    recall  f1-score   support

           alt.atheism       0.97      0.59      0.73       319
         comp.graphics       0.96      0.87      0.91       389
    rec.sport.baseball       0.95      0.97      0.96       397
               sci.med       0.97      0.81      0.88       396
soc.religion.christian       0.64      0.99      0.77       398

              accuracy                           0.86      1899
             macro avg       0.90      0.85      0.85      1899
          weighted avg       0.89      0.86      0.86      1899



# n-Gram

Well, we are kind of stuck at 0.95 f1 score. As we discussed, the bag of words has limitation without knowing the order of words and context in the text. n-Gram contruct a token with a sequence of multiple words to resolve the limitation. Let us practice n-Gram to see if we can further improve the performance.


<div id="TODO9"></div>

### TODO9 (20 points)

For n-Gram, we first need to decide what n to use. So, we start this practice first searching for the hyperparameters. As we naively selected the alpha for Naive Bayes, we search for it as well using `GridSearchCV`.

The hyperparameter dictionary `parameters` needs to follow the naming convention. The name of an object instance in a pipeline followed by *\"__\"* and the hyperparameter name for each instance. For instance, in the code below, for classifier named 'clf', we want to search for alpha. Thus, the dictionary key is *`clf__alpha`*.

1. Finish the parameters with the list of n-gram ranges and classifier alpha values.
2. Create a pipelined classifier with `CountVectorizer`, `TfidfTransformer`, and `MultinomialNB` and name it `ng_clf`.
2. Create a `GridSearchCV` instance with 5 fold cross validation, name it as `gs_clf`, and train it using train data. Use data and target attributes wherever needed.
3. Print `best_score_` of `gs_clf` object.
4. Print the best parameters in `best_params_`.
5. Print out the `classification_report` with `test_data`.


In [17]:
from sklearn.model_selection import GridSearchCV
# 1)
parameters = {
     'vect__ngram_range': [(1, 1), (1, 2), (2, 2)],
     'clf__alpha':[0.1, 1.0, 3.0, 10.0] ,
}


#2)

ng_clf = Pipeline([
    ('vect', CountVectorizer()),
    ('tfidf', TfidfTransformer()),
    ('clf', MultinomialNB()),
])



In [18]:
# 3)
gs_clf = GridSearchCV(ng_clf, parameters, cv=5)


<font color="red">**WARNING: This cell will take a long time to run! If you have more parameters, it will takes longer.**</font>

In [19]:
# 3) cont'
gs_clf.fit(train_data.data, train_data.target)



GridSearchCV(cv=5,
             estimator=Pipeline(steps=[('vect', CountVectorizer()),
                                       ('tfidf', TfidfTransformer()),
                                       ('clf', MultinomialNB())]),
             param_grid={'clf__alpha': [0.1, 1.0, 3.0, 10.0],
                         'vect__ngram_range': [(1, 1), (1, 2), (2, 2)]})

In [20]:
# 4)
gs_clf.best_score_




0.9709171352198359

In [21]:
# 5)
gs_clf.best_params_

{'clf__alpha': 0.1, 'vect__ngram_range': (1, 1)}

In [22]:
# 6)
pred = gs_clf.predict(test_data.data)
print(classification_report(test_data.target, pred, target_names=test_data.target_names))



                        precision    recall  f1-score   support

           alt.atheism       0.97      0.82      0.89       319
         comp.graphics       0.95      0.95      0.95       389
    rec.sport.baseball       0.96      0.99      0.98       397
               sci.med       0.97      0.90      0.93       396
soc.religion.christian       0.84      0.97      0.90       398

              accuracy                           0.93      1899
             macro avg       0.94      0.93      0.93      1899
          weighted avg       0.94      0.93      0.93      1899



# Word Embeddings

Word embeddings are numerical vectors that are corresponding to the meaning and the context of words. In the lecture, we discussed the pretrained Skip Gram word embeddings and RoBERTa in spaCy. By using them, we will experiment the embeddings to the new data and check the classification accuracies.




## Word2Vec

Now, let us use the pretrained language model `en_core_web_lg` in spaCy to test Word2Vec embedding.

<div id="TODO10"></div>

### TODO10 (18 points)


1. Import `spacy`.
2. Load `en_core_web_lg` and store the language model to `nlp`.
   * You can use the line ```spacy.cli.download("en_core_web_lg")``` to download the model.
2. Convert the train and test data into numeric 2D array. Codes for this are provided. Please read them and ask questions if you don't understand the list compresension. This part of codes takes long time, so try to avoid running it multiple times.

3. Create a `LogisticRegression` classifier and train it.
  * Note: logistic regression needs a scaler to normalize input.
4. Make predictions and print the `classification_report`.



In [23]:
# 1)
import spacy

# 2)
spacy.cli.download("en_core_web_lg")
nlp = spacy.load("en_core_web_lg")

✔ Download and installation successful
You can now load the package via spacy.load('en_core_web_lg')
⚠ Restart to reload dependencies
If you are in a Jupyter or Colab notebook, you may need to restart Python in
order to load all the package's dependencies. You can do this by selecting the
'Restart kernel' or 'Restart runtime' option.


<font color="red">**WARNING: This cell will take a long time to run! In my laptop, it took 4 and 2.5 minutes each as you can see in the time report outputs.** </font>

In [24]:
%%time
X_train_lg = np.vstack([nlp(s).vector for s in train_data.data])

CPU times: user 3min 5s, sys: 1.59 s, total: 3min 7s
Wall time: 3min 6s


In [25]:
%%time
X_test_lg = np.vstack([nlp(s).vector for s in test_data.data])

CPU times: user 2min 13s, sys: 1.45 s, total: 2min 14s
Wall time: 2min 14s


In [26]:
X_train_lg.shape, X_test_lg.shape

((2854, 300), (1899, 300))

In [27]:
# 4)
from sklearn.linear_model import LogisticRegression
from sklearn.preprocessing import StandardScaler
from sklearn.pipeline import make_pipeline

lr_clf = make_pipeline(StandardScaler(), LogisticRegression(max_iter=1000))



# 5)
lr_clf.fit(X_train_lg, train_data.target)
predictions = lr_clf.predict(X_test_lg)
print(classification_report(test_data.target, predictions, target_names=test_data.target_names))

                        precision    recall  f1-score   support

           alt.atheism       0.75      0.75      0.75       319
         comp.graphics       0.90      0.90      0.90       389
    rec.sport.baseball       0.93      0.95      0.94       397
               sci.med       0.89      0.83      0.86       396
soc.religion.christian       0.83      0.87      0.85       398

              accuracy                           0.87      1899
             macro avg       0.86      0.86      0.86      1899
          weighted avg       0.87      0.87      0.87      1899



## RoBERTa

You can repeat the TODO10 with `en_core_web_trf` model. This will take long time to run (I failed to run in my laptop!), so if you are interested in running, consider running in faster platform. Therefore, this is optional.

You can refer to this [link](https://huggingface.co/docs/transformers/en/model_doc/roberta) to know more about RoBERTa.

1. Load `en_core_web_trf` and store the language model to `nlp`. You can refer to this [link](https://github.com/explosion/spacy-models/releases) to know more about the package.
2. Convert the train and test data into numeric 2D array. Codes for this are provided. This part of codes takes long time, so try to avoid running it multiple times.

3. Create a `LogisticRegression` classifier and train it.
  * Note: logistic regression needs a scaler to normalize input.
4. Report the `classification_report`.

In [28]:
# 1)
!pip install -U spacy
!python -m spacy info


============================== Info about spaCy ==============================

spaCy version    3.7.4                         
Location         /usr/local/lib/python3.10/dist-packages/spacy
Platform         Linux-6.1.58+-x86_64-with-glibc2.35
Python version   3.10.12                       
Pipelines        en_core_web_lg (3.7.1), en_core_web_sm (3.7.1)



In [29]:
!python -m spacy download en_core_web_trf

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 457.4/457.4 MB 3.1 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 236.3/236.3 kB 5.2 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 731.6/731.6 kB 13.4 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 23.7/23.7 MB 52.0 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 823.6/823.6 kB 52.9 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 14.1/14.1 MB 90.8 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 731.7/731.7 MB 2.7 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 410.6/410.6 MB 3.2 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 121.6/121.6 MB 14.0 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 56.5/56.5 MB 28.4 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 124.2/124.2 MB 8.0 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 196.0/196.0 MB 4.0 MB/s eta 0:00:00
     ━━━━━━━━━━━

In [33]:
import spacy
nlp = spacy.load("en_core_web_trf")

ValueError: [E002] Can't find factory for 'curated_transformer' for language English (en). This usually happens when spaCy calls `nlp.create_pipe` with a custom component name that's not registered on the current language class. If you're using a custom component, make sure you've added the decorator `@Language.component` (for function components) or `@Language.factory` (for class components).

Available factories: attribute_ruler, tok2vec, merge_noun_chunks, merge_entities, merge_subtokens, token_splitter, doc_cleaner, parser, beam_parser, lemmatizer, trainable_lemmatizer, entity_linker, entity_ruler, tagger, morphologizer, ner, beam_ner, senter, sentencizer, spancat, spancat_singlelabel, span_finder, future_entity_ruler, span_ruler, textcat, textcat_multilabel, en.lemmatizer

In [31]:
# Code for converting train data. Uncomment these below lines to run them
%%time
X_train_trf= np.vstack([nlp(s)._.trf_data.tensors[1] for s in train_data.data])

AttributeError: [E046] Can't retrieve unregistered extension attribute 'trf_data'. Did you forget to call the `set_extension` method?

In [32]:
# Code for converting test data. Uncomment these below lines to run them
%%time
X_test_trf = np.vstack([nlp(s)._.trf_data.tensors[1] for s in test_data.data])

AttributeError: [E046] Can't retrieve unregistered extension attribute 'trf_data'. Did you forget to call the `set_extension` method?

In [34]:
# 3)
lr_clf_trf = make_pipeline(StandardScaler(), LogisticRegression(max_iter=1000))

# 4)
lr_clf_trf.fit(X_train_trf, train_data.target)
predictions = lr_clf_trf.predict(X_test_trf)
print(classification_report(test_data.target, predictions, target_names=test_data.target_names))


NameError: name 'X_train_trf' is not defined

# Comparison with Bag-of-Words or n-Gram

Previously, we use `MultinomialNB` for classification but for the word embedding, we used `LogisticRegression`.
As we used different classifiers, it is hard to tell which feature representation works better for the newsgroup dataset. Now, let us repeat the previous work with BOW and n-Gram with the linear logistic regression.

<div id="TODO11"></div>

### TODO11 (20 points)

1. Choose one of the prior TODO practices (TODO2-9) for your choice of features.
2. Replace the `MultinomialNB` with `LogisticRegression`.
  * Note: logistic regression needs a scaler to normalize input.
3. Using `GridSearchCV`, find the best performing model and present the `classificaiont_report`.
4. Discuss your observation of the comparison.

HINT: Referring TODO9, find the best performing parameters and observe the best score.

In [35]:
# parameters to test
parameters = {
    'vect__ngram_range': [(1, 1), (1, 2), (2, 2)],
    'clf__C': [0.01, 0.1, 1, 10, 100],
}


# classifier pipeline
pipeline = Pipeline([
    ('vect', CountVectorizer(min_df=2)),
    ('scaler', StandardScaler(with_mean=False)),
    ('clf', LogisticRegression(max_iter=1000)),
])

In [36]:
# grid search object
gs_clf = GridSearchCV(pipeline, parameters, cv=5, n_jobs=-1)


<font color="red">**WARNING: This cell will take a long time to run! If you have more parameters, it will takes longer.**</font>

In [37]:
# fit
gs_clf.fit(train_data.data, train_data.target)


GridSearchCV(cv=5,
             estimator=Pipeline(steps=[('vect', CountVectorizer(min_df=2)),
                                       ('scaler',
                                        StandardScaler(with_mean=False)),
                                       ('clf',
                                        LogisticRegression(max_iter=1000))]),
             n_jobs=-1,
             param_grid={'clf__C': [0.01, 0.1, 1, 10, 100],
                         'vect__ngram_range': [(1, 1), (1, 2), (2, 2)]})

In [38]:
# test evaluation
predictions = gs_clf.predict(test_data.data)
print(classification_report(test_data.target, predictions, target_names=test_data.target_names))
# Best score and parameters
print("Best Score:", gs_clf.best_score_)
print("Best Parameters:", gs_clf.best_params_)

                        precision    recall  f1-score   support

           alt.atheism       0.98      0.80      0.88       319
         comp.graphics       0.73      0.93      0.82       389
    rec.sport.baseball       0.92      0.95      0.94       397
               sci.med       0.92      0.72      0.81       396
soc.religion.christian       0.86      0.93      0.90       398

              accuracy                           0.87      1899
             macro avg       0.88      0.87      0.87      1899
          weighted avg       0.88      0.87      0.87      1899

Best Score: 0.9446376010077733
Best Parameters: {'clf__C': 0.01, 'vect__ngram_range': (1, 1)}


`TODO 11.4` Is the performan with logistic regression different? What do you think about the embeddings? Do you think it is better than BOW or n-Gram encoding?

`ANSWER`





The performance with Logistic Regression shows a strong outcome with an overall best score of approximately 0.9446 from the GridSearchCV's hyperparameter tuning. This indicates a highly effective model, particularly with the use of unigrams as the best n-gram range. When we consider this in the context of feature representation, it seems that, for this dataset, the simplest form of text representation (unigrams) coupled with Logistic Regression's ability to model complex decision boundaries provides a very competitive approach.


As for embeddings, while they were not directly compared in this output, they typically capture more nuanced semantic information than BOW or n-Gram encoding. This means that while BOW or n-Grams can yield strong performance, as seen here, embeddings could potentially offer even more robust features, especially for tasks requiring understanding of context and semantics.


Comparatively, the choice between using embeddings versus BOW/n-Grams might depend on the complexity of the task and the nature of the data. BOW and n-Grams are often faster to compute and can perform exceptionally well in many cases, but embeddings can provide superior performance when you need to capture semantic meaning and use models that can leverage the dense representation of embeddings effectively. However, the fact that Logistic Regression with a basic unigram representation has performed so well here suggests that for some datasets and problems, the additional complexity of embeddings might not be necessary.


<div id="Feedback"></div>

## Feedback (2 point)

Did you enjoy the lab?

Please take time to answer the following feedback qustions to help us further improve these labs! Your feedback is crucial to making these labs more useful!
    


* How do you rate the overall experience in this lab? (5 likert scale. i.e., 1 - poor ... 5 - amazing)  
Why do you think so? What was most/least useful?



`ANSWER`

The overall experience in this lab is 5 out of 5. All the concepts are useful in this lab.

* What did you find difficult about the lab? Were there any TODOs that were unclear? If so, what specfically did not make sense about it?



`ANSWER`

Todos are clear and easy to follow.

* Which concepts, if any, within the lab do you feel could use more explanation?

`ANSWER`

All the concepts are clear and understable.